pkg/helm/chartproxy/repos_test.go

```diff
@@ -262,11 +263,12 @@ func TestHelmRepoGetter_ListErrors(t *testing.T) {
 
 func TestHelmRepo_IndexFile(t *testing.T) {
 	tests := []struct {
-		name      string
-		url       string
-		httpCode  int
-		indexFile string
-		err       bool
+		name          string
+		url           string
+		httpCode      int
+		indexFile     string
+		resultDiffers bool
```

Comment:
so far we were assuming that `indexFile` is the expected one, but that might not be true anymore. Instead of introducing just a flag, and for future improvements, how about that we introduce `expectedIndexFile` field. If set, it is going to be used, otherwise, we continue using `indexFile` as backup in the test assertion part.

Reply:
While I agree using an `expectedIndexFile` would be the preferred way, I can't think of any sane way to compare both files. Since the `httptest.NewServer` randomly allocates an address:port pair that is assigned to `connectionConfig.url` and this commit prepends that url to the relative chart repository url's we can't reliably determine how the `expectedIndexFile` should look like. Any suggestions are welcome!
https://github.com/openshift/console/blob/ccd330f0fcb060cafcfab622654ca253f87b9bb8/pkg/helm/chartproxy/repos_test.go#L317-L317
https://github.com/openshift/console/blob/ccd330f0fcb060cafcfab622654ca253f87b9bb8/pkg/helm/chartproxy/repos_test.go#L329-L329
https://github.com/openshift/console/blob/ccd330f0fcb060cafcfab622654ca253f87b9bb8/pkg/helm/chartproxy/repos.go#L94

Reply:
Good point! Let's then drop usage http server. In order to do that we need to slightly refactory `httpRepo` struct:

```diff
diff --git a/pkg/helm/chartproxy/repos.go b/pkg/helm/chartproxy/repos.go
index 89e7548b6..5232447e1 100644
--- a/pkg/helm/chartproxy/repos.go
+++ b/pkg/helm/chartproxy/repos.go
@@ -36,25 +36,13 @@ const (
 )
 
 type helmRepo struct {
-	Name            string
-	URL             *url.URL
-	Disabled        bool
-	TLSClientConfig *tls.Config
+	Name       string
+	URL        *url.URL
+	Disabled   bool
+	httpClient func() (*http.Client, error)
 }
 
-type TLSConfigGetter interface {
-	Get() (*tls.Config, error)
-}
-
-func (repo helmRepo) Get() (*tls.Config, error) {
-	return repo.TLSClientConfig, nil
-}
-
-func (repo helmRepo) httpClient() (*http.Client, error) {
-	tlsConfig, err := repo.Get()
-	if err != nil {
-		return nil, err
-	}
+func httpClient(tlsConfig *tls.Config) (*http.Client, error) {
 	tr := &http.Transport{
 		TLSClientConfig: tlsConfig,
 	}
@@ -169,7 +157,7 @@ func (b helmRepoGetter) unmarshallConfig(repo unstructured.Unstructured) (*helmR
 			return nil, err
 		}
 	}
-	h.TLSClientConfig = crypto.SecureTLSConfig(&tls.Config{
+	tlsClientConfig := crypto.SecureTLSConfig(&tls.Config{
 		RootCAs: rootCAs,
 	})
 	if tlsReference != "" {
@@ -192,9 +180,12 @@ func (b helmRepoGetter) unmarshallConfig(repo unstructured.Unstructured) (*helmR
 			if err != nil {
 				return nil, err
 			}
-			h.TLSClientConfig.Certificates = []tls.Certificate{cert}
+			tlsClientConfig.Certificates = []tls.Certificate{cert}
 		}
 	}
+	h.httpClient = func() (*http.Client, error) {
+		return httpClient(tlsClientConfig)
+	}
 	return h, nil
 }
 
diff --git a/pkg/helm/chartproxy/repos_test.go b/pkg/helm/chartproxy/repos_test.go
index 80f1274d1..8b170a4dd 100644
--- a/pkg/helm/chartproxy/repos_test.go
+++ b/pkg/helm/chartproxy/repos_test.go
@@ -2,11 +2,10 @@ package chartproxy
 
 import (
 	"errors"
-	"fmt"
 	"io/ioutil"
 	"net/http"
-	"net/http/httptest"
 	neturl "net/url"
+	"os"
 	"reflect"
 	"testing"
 
@@ -26,6 +25,12 @@ type apiError struct {
 	msg      string
 }
 
+type RoundTripFunc func(req *http.Request) *http.Response
+
+func (f RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req), nil
+}
+
 func onlyResult(obj interface{}, err error) interface{} {
 	return obj
 }
@@ -301,21 +306,10 @@ func TestHelmRepo_IndexFile(t *testing.T) {
 				Client: fake.K8sDynamicClient(),
 			}
 			url := tt.url
-
 			if url == "" {
-				ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-					w.Header().Set("Content-Type", "application/yaml")
-					w.WriteHeader(tt.httpCode)
-					if tt.indexFile != "" {
-						content, err := ioutil.ReadFile(tt.indexFile)
-						if err != nil {
-							t.Error(err)
-						}
-						fmt.Fprintln(w, string(content))
-					}
-				}))
-				url = ts.URL
+				url = "http://repo.foo.com"
 			}
+
 			repoCR := unstructured.Unstructured{
 				Object: map[string]interface{}{
 					"apiVersion": "helm.openshift.io/v1beta1",
@@ -335,7 +329,27 @@ func TestHelmRepo_IndexFile(t *testing.T) {
 			if err != nil {
 				t.Error(err)
 			}
+			if tt.url == "" {
+				repo.httpClient = func() (*http.Client, error) {
+					return &http.Client{
+						Transport: RoundTripFunc(func(req *http.Request) *http.Response {
+							resp := &http.Response{
+								StatusCode: tt.httpCode,
+							}
+							if tt.indexFile != "" {
+								r, err := os.Open(tt.indexFile)
+								if err != nil {
+									t.Error(err)
+								}
+								resp.Body = ioutil.NopCloser(r)
+							}
+							return resp
+						}),
+					}, nil
+				}
+			}
 			index, err := repo.IndexFile()
+
 			if tt.err && err == nil {
 				t.Errorf("Expected error %v but got %v", tt.err, err)
 			}

```

with the above, we can fully then mock http client and have fix repo url all the time.

Reply:
Thanks for the very detailed suggestion!